docs/proposals/20190708-MachineConfig.md

```diff
@@ -0,0 +1,206 @@
+    endpoints: [ {{ .Endpoints }} ]
+    certSANs: [ "{{ .IP }}", "127.0.0.1" ]
+networking:
+  os:
+    devices:
+      - interface: lo
+        cidr: x.x.x.x/32
+      - interface: eth0
+        dhcp: true
+install:
+  wipe: false
+  force: true
+  boot:
+    device: /dev/sda
+    size: 1024000000
+    kernel: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/vmlinuz
+    initramfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/initramfs.xz
+  root:
+    device: /dev/sda
+    size: 2048000000
+    rootfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/rootfs.tar.gz
+  data:
+    device: /dev/sda
+    size: 4096000000
+```
+
+## Proposal
+
+There are quite a few fields that are similar among all node types, while the main difference between them is the kubeadm templates. Given all the proper info, we should be able to generate the kubeadm configuration on the fly depending on node type and abstract the need to know all the different options away. I'm proposing a machine configuration that looks like this:
+
+```
+#!talos
+## We'll implement a versioning strategy and 
+## respect it when parsing the config
+version: "v0.1.0"
+## All security focused fields (certs and tokens for trustd/kubeadm)
+## go under the security key
+security:
+  os:
+    ca:
+      crt: "{{ .Certs.OsCert }}"
+      key: "{{ .Certs.OsKey }}"
+    token: abc.1234
+  kubernetes:
+    ca:
+      crt: "{{ .Certs.K8sCert }}"
+      key: "{{ .Certs.K8sKey }}"
+    token: abc.1234
+## All configs that are specific to this machine only
+## go in machine key
+machine:
+  nodeType: init, controlplane, worker
+  kubeletExtraArgs: {}
+  network:
+    interfaces:
+      - name: public
+        mac: aa:bb:cc:ee:ff:00
+        addresses:
+          - "10.254.10.11/24"
+          - "192.168.251.11/24"
+          - "2001:db8:1000::54/64"
+      - name: internal
+        device: eth1
+        addresses:
+          - "55.44.11.23/29"
+        ## An example of how we might provide routes for a given interface
+        routes:
+          - "2001:db8::10:0/64 via 2001:db8::0:2bd9 metric 10"
+          - "0.0.0.0/0 via 192.168.24.1"
+          - "10.66.254.0/24 via 10.254.254.1 dev tap0"
+  ## This is currently just an idea of how another talos "service" 
+  ## might get configured for the machine
+  ntp:
+    servers:
+      - server1
+## All cluster-wide configs go here. These are generally things that get 
+## injected to kubeadm, or affect the behavior of the kubernetes control 
+## plane as a whole.
+cluster:
+  clusterName: {{ .ClusterName }}
+  controlPlane:
+    ips: [x.x.x.x, y.y.y.y, z.z.z.z]
+    apiServerExtraArgs: {}
+    controllerManagerExtraArgs: {}
+    schedulerExtraArgs: {}
+    network:
+      dnsDomain: {{ .ServiceDomain }}
+      podSubnet: {{ index .PodNet 0 }}
+      serviceSubnet: {{ index .ServiceNet 0 }}
+    ## An example of how we might allow for override of image locations with local registries
```

Comment:
Exposed the images we'll use for control plane to allow folks to use their own repos or different versions (though we may want to say that's not really supported but may work)